src/Components/test/testassets/BasicTestApp/ErrorBoundaryTest/ErrorBoundaryCases.razor

```diff
@@ -143,4 +151,39 @@
         // Before it completes, dispose its enclosing error boundary
         disposalTestRemoveErrorBoundary = true;
     }
+
+    async Task SyncExceptionDispatch()
+    {
+        try
+        {
+            await InvokeAsync(() =>
+            {
+                throw new InvalidTimeZoneException("Synchronous exception in SyncExceptionDispatch");
+            });
+        }
+        catch (Exception e)
+        {
+            await DispatchExceptionAsync(e);
+        }
+    }
+
+    async Task AsyncExceptionDispatch()
+    {
+        try
+        {
+            await InvokeAsync(async () =>
+            {
+                await ThrowExceptionAsync();
+            });
+        }
+        catch (Exception e)
+        {
+            await DispatchExceptionAsync(e);
+        }
+    }
+
+    private Task ThrowExceptionAsync()
```

Comment:
This is not throwing the exception asynchronously. The thread is not yielded at any point, so this is similar/equivalent to the `SyncExceptionDispatch` case.

For the exception to be thrown asynchronously you need to use something like `await Task.Yield()` before throwing the exception.

Also, I will point out that we normally also write unit tests for these types of behavior in the Renderer itself. There we normally use a TaskCompletionSource to very accurately control the process.

Reply:
Good point.

In fact more generally I think these test cases are missing the key scenario: when an exception happens *outside* the normal lifecycle flow and you couldn't just do a normal `throw` anyway. With the way `SyncExceptionDispatch` and `AsyncExceptionDispatch` are structured, `DispatchExceptionAsync` is no different from simply using `throw` at that point, because we're still within the task chain of the lifecycle method.

It would be good to cover the primary use case where this new feature is what causes the exception to be associated with the renderer. For example, by using `Task.Run` to simulate having some code running completely separately from the renderer callstack/task chain:

```cs
    void ExternalExceptionDispatch()
    {
        Task.Run(() =>
        {
            // Inside Task.Run, we're outside the call stack or task chain of the lifecycle method, so
            // DispatchExceptionAsync is needed to get an exception back into the component
            _ = DispatchExceptionAsync(new InvalidTimeZoneException());
        });
    }
```

If we are going to have the `SyncExceptionDispatch`/`AsyncExceptionDispatch` cases, I imagine they would be there to cover the case where people are calling `DispatchExceptionAsync` even though they didn't need to (i.e., calls that are already on the lifecycle method call stack), so they could be simplified to just dispatching a newed-up exception like:

```cs
    async Task SyncExceptionDispatch()
    {
        await DispatchExceptionAsync(new InvalidTimeZoneException("Sync"));
    }

    async Task AsyncExceptionDispatch()
    {
        await Task.Yield();
        await DispatchExceptionAsync(new InvalidTimeZoneException("Async"));
    }
```